docs/design/2018-10-19-table-partition.md

```diff
@@ -0,0 +1,147 @@
+
+Table partition works on the first level of mapping, partition ID will be made equivalent of the table ID. A partitioned table row uses `partition id + row id` as encoded key.
+
+Table ID and partition IDs relationship should be maintained in the `TableInfo`.
+
+When a new row is inserted into a partitioned table, if the row doesn't  belong to any partition of the table, the operation fails.
+
+Handling `NULL` values refers to MySQL document https://dev.mysql.com/doc/refman/8.0/en/partitioning-handling-nulls.html
+
+### How to read from the partitioned table
+
+```
+create table t (id int) partition by range (id)
+(partition p1 values less than (10),
+partition p2 values less than (20),
+partition p3 values less than (30))
+```
+
+This SQL query
+
+```
+select * from t 
+```
+
+is equal to
+
+```
+select * from (union all
+select * from p1 where id < 10
+select * from p2 where id < 20
+select * from p3 where id < 30)
+```
+
+During logical optimize phase, the `DataSource` plan is translated into `UnionAll`, then each partition generates its own `TableReader` in the physical optimize phase.
+
+The drawbacks of this implementation are:
+
+* If the table has many partitions, there will be many readers, then `explain` result is not friendly to user
+* `UnionAll` executor can't keep order, so if some executor need ordered result such as `IndexReader`, an extra `Sort` executor is needed
+
+If [partition selection](https://dev.mysql.com/doc/refman/5.7/en/partitioning-selection.html) is used, planner should translate table ID to partition ID, and turn off partition pruning.
+
+#### Partition pruning
+
+During logical optimize phase, after predicate push down, partition pruning is performed.
+
+For range partition, partition pruning is based on the range filter conditions of the partition column. For hash partition, if the filter condition is something like `key = const`, partition pruning could work.
+
+### How to write to the partitioned table
+
+All the write operation calls functions like `table.AddRecord` eventually, so implement write operation on a partitioned table simply implement this interface method on the `PartitionedTable` struct.
+
+`PartitionedTable` implements the `table.Table` interface, overloads the `AddRecord` method. It also comes with a `locatePartition` method, to decide which partition a row should be insert into.
+
+Each partition maintains its own index data, the insert operation should keep data and index consistent.
+
+### DDL operations
+
+DROP/TRUNCATE/ADD, those three operations works on range partition.
+
+`DROP PARTITION` is similar to `DROP TABLE`, except that partition ID is used. `TableInfo` should be updated after the drop partition operation. Note that if one want to drop the last partition in a table, he should use drop table rather than drop partition.
```

Comment:
`DROP PARTITION` is similar to `DROP TABLE`, except that partition ID is used. `TableInfo` should be updated after the drop partition operation. Note that if one **wants** to drop the last partition in a table, he should use drop table rather than drop partition.